workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt

```diff
@@ -192,16 +234,80 @@ namespace <#= qualifiedNamespace #>
                         continue;
                     }
 
-                    var authChanges = authChangeLists[i];
-                    foreach (Requirables.ReaderWriterImpl readerWriter in readersWriters)
+                    if (IsFirstAuthChange(Authority.Authoritative, changeOpsLists[i]))
                     {
-                        foreach (var auth in authChanges.Changes)
+                        foreach (Requirables.ReaderWriterImpl readerWriter in readersWriters)
                         {
-                            readerWriter.OnAuthorityChange(auth);
+                            readerWriter.OnAuthorityChange(Authority.Authoritative);
                         }
                     }
                 }
             }
+
+            public override void InvokeOnAuthorityLostCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore)
+            {
+                if (AuthorityLostComponentGroup.IsEmptyIgnoreFilter)
+                {
+                    return;
+                }
+
+                var entities = AuthorityLostComponentGroup.GetEntityArray();
+                var changeOpsLists = AuthorityLostComponentGroup.GetComponentDataArray<AuthorityChanges<<#= componentDetails.TypeName #>>>();
+
+                // Call once on all entities unless they flip-flopped back into the state they started in
+                for (var i = 0; i < entities.Length; i++)
+                {
+                    var injectableStore = entityIndexToInjectableStore[entities[i].Index];
+                    if (!injectableStore.TryGetInjectablesForComponent(readerWriterInjectableId, out var readersWriters))
+                    {
+                        continue;
+                    }
+
+                    if (IsFirstAuthChange(Authority.NotAuthoritative, changeOpsLists[i]))
+                    {
+                        foreach (Requirables.ReaderWriterImpl readerWriter in readersWriters)
+                        {
+                            readerWriter.OnAuthorityChange(Authority.NotAuthoritative);
+                        }
+                    }
+                }
+            }
+
+            private bool IsFirstAuthChange(Authority authToMatch, AuthorityChanges<<#= componentDetails.TypeName #>> changeOps)
+            {
+                foreach (var auth in changeOps.Changes)
+                {
+                    if (auth != Authority.AuthorityLossImminent) // not relevant
+                    {
+                        return auth == authToMatch;
+                    }
+                }
+                return false;
+            }
+
+            public override void InvokeOnAuthorityLossImminentCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore)
```

Comment:
Nit: Could you place this function below the `InvokeOnAuthorityLostCallbacks()` function? Basically flip the order of this function an `IsFirstAuthChange()`. That way, similarly functions are grouped together.